### PR TITLE
feat: refactor estimate worker model needs & add spawn infos

### PR DESCRIPTION
**Pull request by yesnault**
**DO NOT MERGE YET**

PR just for Information.

added on pipelineBuildJob:

```json
"spawninfos": [
            {
              "api_time": "2017-03-12T17:20:27.367833327+01:00",
              "remote_time": "2017-03-12T17:20:27.347129755+01:00",
              "info": "Hatchery (754) starts spawn worker with model laptop712297.home-hopeful_jones"
            },
            {
              "api_time": "2017-03-12T17:20:27.367833327+01:00",
              "remote_time": "2017-03-12T17:20:27.359649813+01:00",
              "info": "Hatchery (754) spawn worker successfully in 0s"
            },
            {
              "api_time": "2017-03-12T17:20:52.018298311+01:00",
              "remote_time": "2017-03-12T17:20:52.010104334+01:00",
              "info": "Action taken by worker laptop712297.home-hopeful_jones-sad_williams"
            },
            {
              "api_time": "2017-03-12T17:20:52.018298311+01:00",
              "remote_time": "2017-03-12T17:20:52.010104334+01:00",
              "info": "This worker was created to take this action"
            },
            {
              "api_time": "2017-03-12T17:21:25.146016381+01:00",
              "remote_time": "2017-03-12T17:21:22.124697383+01:00",
              "info": "worker laptop712297.home-hopeful_jones-sad_williams finished working on this job and took 30s to work on the steps"
            }
          ]
```

#333

### CODE REVIEW

**Comment by fsamin** [comment]
cool, maybe info should be structured data to be able to translate it (see package engine/api/msg) on call made by the client

**Comment by yesnault** [comment]
before: 
```go
type SpawnInfo struct {
	APITime    time.Time `json:"api_time,omitempty" db:"-"`
	RemoteTime time.Time `json:"remote_time,omitempty" db:"-"`
	Info       string    `json:"info,omitempty" db:"-"`
}
```

after:
```go
// SpawnInfo contains an information about spawning
type SpawnInfo struct {
	APITime    time.Time `json:"api_time,omitempty" db:"-"`
	RemoteTime time.Time `json:"remote_time,omitempty" db:"-"`
	Message    SpawnMsg  `json:"message,omitempty" db:"-"`
}

// SpawnMsg represents a msg for spawnInfo
type SpawnMsg struct {
	// Key is defined under github.com/api/engine/msg
	Key    string   `json:"key" db:"-"`
	Values []string `json:"values" db:"-"`
}
```

**Comment by yesnault** [comment]
it's now:

```json
"spawninfos": [
  {
    "api_time": "2017-03-12T22:22:53.726524554+01:00",
    "remote_time": "2017-03-12T22:22:53.719686886+01:00",
    "message": {
      "id": 1,
      "args": [
        "756",
        "laptop712297.home-small_pike"
      ]
    },
    "user_message": "Hatchery (756) starts spawn worker with model laptop712297.home-small_pike"
  },
  {
    "api_time": "2017-03-12T22:22:53.726524554+01:00",
    "remote_time": "2017-03-12T22:22:53.723335241+01:00",
    "message": {
      "id": 3,
      "args": [
        "756",
        "0s"
      ]
    },
    "user_message": "Hatchery (756) spawn worker successfully in 0s"
  },
  {
    "api_time": "2017-03-12T22:23:04.729297748+01:00",
    "remote_time": "2017-03-12T22:23:04.724259963+01:00",
    "message": {
      "id": 5,
      "args": [
        "laptop712297.home-small_pike-silly_bhabha"
      ]
    },
    "user_message": "Job was taen by worker laptop712297.home-small_pike-silly_bhabha"
  },
  {
    "api_time": "2017-03-12T22:23:04.729297748+01:00",
    "remote_time": "2017-03-12T22:23:04.724259963+01:00",
    "message": {
      "id": 6,
      "args": [
        "laptop712297.home-small_pike-silly_bhabha"
      ]
    },
    "user_message": "This worker was created to take this action%!(EXTRA string=laptop712297.home-small_pike-silly_bhabha)"
  },
  {
    "api_time": "2017-03-12T22:23:37.813471695+01:00",
    "remote_time": "2017-03-12T22:23:34.799800126+01:00",
    "message": {
      "id": 4,
      "args": [
        "laptop712297.home-small_pike-silly_bhabha",
        "30s"
      ]
    },
    "user_message": "Worker laptop712297.home-small_pike-silly_bhabha finished working on this job and took 30s to work on the steps"
  }
]
```

**Comment by yesnault** [comment]
with https://github.com/ovh/cds/pull/350/commits/fb4a71835765c2e2e6edf980f4125201c6a30573#diff-f1f73df04aa151b9581bb9da16e67dff

File: engine/api/build.go [hunk 1]
```diff
@@ -408,8 +432,72 @@ func takeActionBuildHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbM
 	return WriteJSON(w, r, pbji, http.StatusOK)
 }
 
-func loadActionBuildSecrets(db *gorp.DbMap, pbJobID int64) ([]sdk.Variable, error) {
+func preCheckHatchery(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *context.Ctx) (int64, error) {
```
Comment (sguiheux) [inline engine/api/build.go]: Ca serait pus propre de definir une fonction niveau handler pour restreindre à une hatchery, (cf Route pour admin)

File: engine/api/action.go [hunk 2]
```diff
@@ -297,17 +297,17 @@ func importActionHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap,
 		return sdk.ErrWrongRequest
```
Comment (sguiheux) [inline engine/api/action.go]: return err

Comment (yesnault) [inline engine/api/action.go]: nope. return err just before

File: engine/api/pipeline/awol.go [hunk 3]
```diff
@@ -45,9 +45,9 @@ func AWOLPipelineKiller(DBFunc func() *gorp.DbMap) {
 func killAWOLPipelineBuildJob(db *gorp.DbMap, pbJobData awolPipelineBuildJob) error {
 	log.Warning("killAWOLPipelineBuildJob> Killing pipeline_job_build %d\n", pbJobData.pipelineBuildJobID)
 
-	tx, err := db.Begin()
-	if err != nil {
-		return err
+	tx, errb := db.Begin()
+	if errb != nil {
+		return errb
```
Comment (fsamin) [inline engine/api/pipeline/awol.go]: wrap error

File: engine/api/pipeline/awol.go [hunk 4]
```diff
@@ -62,8 +62,7 @@ func killAWOLPipelineBuildJob(db *gorp.DbMap, pbJobData awolPipelineBuildJob) er
 	}
 
 	query := `UPDATE worker SET status = $1, action_build_id = NULL WHERE action_build_id = $2`
-	_, err = tx.Exec(query, string(sdk.StatusDisabled), pbJobData.pipelineBuildJobID)
-	if err != nil {
+	if _, err := tx.Exec(query, string(sdk.StatusDisabled), pbJobData.pipelineBuildJobID); err != nil {
 		return err
 	}
```
Comment (fsamin) [inline engine/api/pipeline/awol.go]: wrap error

File: engine/api/pipeline/gorp_model.go [hunk 5]
```diff
@@ -35,12 +36,21 @@ func (p *PipelineBuildJob) PostInsert(s gorp.SqlExecutor) error {
 
 //PostGet is a DB Hook on PipelineBuildJob to get jobs and params as from JSON in DB
 func (p *PipelineBuildJob) PostGet(s gorp.SqlExecutor) error {
+	h := sdk.Hatchery{}
+	if cache.Get(keyBookJob(p.ID), &h) {
```
Comment (fsamin) [inline engine/api/pipeline/gorp_model.go]: i did not see where it put in cache ?

File: engine/api/router.go [hunk 6]
```diff
@@ -219,50 +220,53 @@ func (r *Router) Handle(uri string, handlers ...RouterConfigParam) {
 		}
 
 		permissionOk := true
-		if rc.auth && rc.needAdmin && !c.User.Admin {
+		if rc.auth && rc.needHatchery && c.Hatchery == nil {
```
Comment (fsamin) [inline engine/api/router.go]: i don't understand :(

File: engine/api/user.go [hunk 7]
```diff
@@ -118,11 +118,10 @@ func UpdateUserHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c
...
 
-	userDB, err := user.LoadUserWithoutAuth(db, username)
-	if err != nil {
-		fmt.Printf("getUserHandler: Cannot load user from db: %s\n", err)
-		return err
-
+	userDB, errload := user.LoadUserWithoutAuth(db, username)
+	if errload != nil {
+		fmt.Printf("getUserHandler: Cannot load user from db: %s\n", errload)
+		return errload
```
Comment (fsamin) [inline engine/api/user.go]: wrap error

File: engine/api/user.go [hunk 8]
```diff
@@ -137,8 +136,7 @@ func UpdateUserHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c
 		return sdk.ErrWrongRequest
 	}
 
-	err = user.UpdateUser(db, userBody)
-	if err != nil {
+	if err := user.UpdateUser(db, userBody); err != nil {
 		log.Warning("updateUserHandler: Cannot update user table: %s", err)
 		return err
```
Comment (fsamin) [inline engine/api/user.go]: wrap error

File: engine/hatchery/swarm/swarm.go [hunk 9]
```diff
@@ -311,7 +307,10 @@ func (h *HatcherySwarm) createAndStartContainer(name, image, network, networkAli
 }
 
 // CanSpawn checks if the model can be spawned by this hatchery
-func (h *HatcherySwarm) CanSpawn(model *sdk.Model, req []sdk.Requirement) bool {
+func (h *HatcherySwarm) CanSpawn(model *sdk.Model, job *sdk.PipelineBuildJob) bool {
+
+	// TODO CHECK RATIO
```
Comment (fsamin) [inline engine/hatchery/swarm/swarm.go]: TODO

File: sdk/hatchery/hatchery.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+					log.Warning("routine> cannot record AddSpawnInfosPipelineBuildJob for job %d: %s\n", job.ID, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func canRunJob(h Interface, job *sdk.PipelineBuildJob, model *sdk.Model, hostname string) bool {
+
```
Comment (fsamin) [inline sdk/hatchery/hatchery.go]: empty line

File: engine/api/pipeline/gorp_model.go (hunk 5 above)
Comment (yesnault) [inline engine/api/pipeline/gorp_model.go]: pipeline_build_job.go

```go
// BookPipelineBuildJob Book an action for a hatchery
func BookPipelineBuildJob(pbJobID int64, hatchery *sdk.Hatchery) (*sdk.Hatchery, error) {
	k := keyBookJob(pbJobID)
	h := sdk.Hatchery{}
	if !cache.Get(k, &h) {
		// job not already booked, book it for 2 min
		cache.SetWithTTL(k, hatchery, 120)
		return nil, nil
	}
	return &h, ErrAlreadyBooked
}
```

File: engine/api/build.go [hunk 11]
```diff
@@ -408,8 +424,67 @@ func takeActionBuildHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbM
...
+
+	if err := tx.Commit(); err != nil {
+		log.Warning("addSpawnInfosPipelineBuildJobHandler> Cannot commit tx: %s\n", err)
+		return sdk.ErrUnknownError
+	}
+
+	return WriteJSON(w, r, nil, http.StatusOK)
+}
+
+func loadActionBuildSecrets(db *gorp.DbMap, pbJobID int64) ([]sdk.Variable, error) {
```
Comment (yesnault) [inline engine/api/build.go]: not possible
```
import cycle not allowed
package github.com/ovh/cds/engine/api
	imports github.com/ovh/cds/engine/api/application
	imports github.com/ovh/cds/engine/api/hook
	imports github.com/ovh/cds/engine/api/pipeline
	imports github.com/ovh/cds/engine/api/project
	imports github.com/ovh/cds/engine/api/application
```

File: engine/api/application/dao.go [hunk 12]
```diff
@@ -238,8 +236,6 @@ func LoadAll(db gorp.SqlExecutor, key string, u *sdk.User, opts ...LoadOptionFun
 }
 
 func loadapplications(db gorp.SqlExecutor, u *sdk.User, opts []LoadOptionFunc, query string, args ...interface{}) ([]sdk.Application, error) {
-	log.Debug("application.load> %s %v", query, args)
```
Comment (yesnault) [inline engine/api/application/dao.go]: fixed

File: engine/api/build.go [hunk 13]
```diff
@@ -408,8 +424,67 @@ func takeActionBuildHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbM
...
+			log.Warning("bookPipelineBuildJobHandler> job %d already booked by %s (%d): %s\n", id, h.Name, h.ID)
+			return WriteJSON(w, r, "job already booked", http.StatusConflict)
+		}
+		log.Warning("bookPipelineBuildJobHandler> Cannot book job %d: %s\n", id, err)
+		return err
+	}
+	return WriteJSON(w, r, nil, http.StatusOK)
+}
+
+func addSpawnInfosPipelineBuildJobHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *context.Ctx) error {
```
Comment (yesnault) [inline engine/api/build.go]: added

File: engine/api/router.go (hunk 6 above)
Comment (yesnault) [inline engine/api/router.go]: c.Hatchery is set in func (r *Router) checkHatcheryAuth(db *gorp.DbMap, headers http.Header, c *context.Ctx) error 

needHatchery is set with func NeedHatchery() RouterConfigParam

NeedHatchery() is added on main_route

File: engine/api/pipeline/gorp_model.go (hunk 5 above)
Comment (fsamin) [inline engine/api/pipeline/gorp_model.go]: ok

File: engine/api/router.go (hunk 6 above)
Comment (fsamin) [inline engine/api/router.go]: ok

File: engine/api/build.go (hunk 11 above)
Comment (fsamin) [inline engine/api/build.go]: :disappointed:

File: engine/hatchery/swarm/swarm.go [hunk 14]
```diff
@@ -29,12 +29,11 @@ type HatcherySwarm struct {
...
 //Init connect the hatchery to the docker api
 func (h *HatcherySwarm) Init() error {
-	var err error
-
-	h.dockerClient, err = docker.NewClientFromEnv()
-	if err != nil {
+	var errc error
+	h.dockerClient, errc = docker.NewClientFromEnv()
+	if errc != nil {
 		log.Critical("Unable to connect to a docker client")
```
Comment (sguiheux) [inline engine/hatchery/swarm/swarm.go]: don't log err ?

File: engine/api/build.go [hunk 15]
```diff
@@ -348,8 +348,8 @@ func takePipelineBuildJobHandler(w http.ResponseWriter, r *http.Request, db *gor
 
 	pbJob, errTake := pipeline.TakePipelineBuildJob(tx, id, workerModel, caller.Name, infos)
 	if errTake != nil {
-		if errTake != pipeline.ErrAlreadyTaken {
-			return sdk.WrapError(errTake, "takePipelineBuildJobHandler> Cannot give ActionBuild %d", id)
+		if errTake != sdk.ErrAlreadyTaken {
+			return sdk.WrapError(errTake, "takePipelineBuildJobHandler> Cannot take job %d", id)
 		}
 		return errTake
```
Comment (sguiheux) [inline engine/api/build.go]: where is the 🌯

File: engine/hatchery/marathon/marathon.go [hunk 16]
```diff
@@ -122,6 +122,7 @@ func (m *HatcheryMarathon) CanSpawn(model *sdk.Model, job *sdk.PipelineBuildJob)
 	}
 	//Service requirement are not supported
 	for _, r := range job.Job.Action.Requirements {
+		log.Notice("CanSpawn> Job %s has a service requirement. Marathon can't spawn a worker for this job %s", job.ID)
```
Comment (sguiheux) [inline engine/hatchery/marathon/marathon.go]: 2 %s,   and one int arg , strange

File: engine/api/build.go [hunk 17]
```diff
@@ -368,36 +334,42 @@ func takeActionBuildHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbM
...
+	}
+
+	pbJob, errTake := pipeline.TakePipelineBuildJob(tx, id, workerModel, caller.Name, infos)
 	if errTake != nil {
-		if errTake != pipeline.ErrAlreadyTaken {
-			log.Warning("takeActionBuildHandler> Cannot give ActionBuild %d: %s\n", id, errTake)
+		if errTake != sdk.ErrAlreadyTaken {
+			return sdk.WrapError(errTake, "takePipelineBuildJobHandler> Cannot take job %d", id)
 		}
 		return errTake
```
Comment (sguiheux) [inline engine/api/build.go]: Where is the 🌯

File: engine/hatchery/marathon/marathon.go [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+
+// CanSpawn return wether or not hatchery can spawn model
+// requirements services are not supported
+func (m *HatcheryMarathon) CanSpawn(model *sdk.Model, job *sdk.PipelineBuildJob) bool {
+	if model.Type != sdk.Docker {
+		return false
+	}
+	//Service requirement are not supported
+	for _, r := range job.Job.Action.Requirements {
+		log.Notice("CanSpawn> Job %d has a service requirement. Marathon can't spawn a worker for this job %s", job.ID)
```
Comment (sguiheux) [inline engine/hatchery/marathon/marathon.go]: missing value for %s

File: engine/hatchery/marathon/marathon.go [hunk 19]
```diff
@@ -0,0 +1,472 @@
...
+	HatcheryID     int64
+	JobID          int64
+	MarathonID     string
+	MarathonVHOST  string
+	MarathonLabels string
+	Memory         int
+	WorkerTTL      int
+}
+
+const marathonPOSTAppTemplate = `
```
Comment (fsamin) [inline engine/hatchery/marathon/marathon.go]: remove this and instantiate a marathon config properly

File: engine/hatchery/main.go [hunk 20]
```diff
@@ -97,7 +96,10 @@ func addFlags() {
 	rootCmd.PersistentFlags().BoolP("insecure", "k", false, `(SSL) This option explicitly allows hatchery to perform "insecure" SSL connections on CDS API.`)
 	viper.BindPFlag("insecure", rootCmd.PersistentFlags().Lookup("insecure"))
 
-	rootCmd.PersistentFlags().BoolP("random-name", "", false, `Generate a random name for hatchery`)
+	rootCmd.PersistentFlags().BoolP("name", "", false, `The name for hatchery <name>-cds-<type>`)
+	viper.BindPFlag("name", rootCmd.PersistentFlags().Lookup("name"))
+
+	rootCmd.PersistentFlags().BoolP("random-name", "", false, `Generate a random name for hatchery <name>-cds-<type>-<random-name>`)
```
Comment (fsamin) [inline engine/hatchery/main.go]: random-name should be true if name is empty, else false

File: engine/hatchery/openstack/openstack.go [hunk 21]
```diff
@@ -804,6 +802,10 @@ func getServers(endpoint, token string) ([]Server, error) {
 		if !worker {
 			continue
 		}
+		workerHatcheryName, ok := s.Metadata["hatcheryName"]
+		if !ok || workerHatcheryName != h.Hatchery().Name {
```
Comment (fsamin) [inline engine/hatchery/openstack/openstack.go]: we should delete server with medata worker and without hatcheryName metadata to be able to delete old server just after put in production

File: engine/hatchery/marathon/marathon.go [hunk 22]
```diff
@@ -364,16 +366,20 @@ func (m *HatcheryMarathon) killAwolWorkers() error {
...
+		log.Debug("killAwolWorkers> check app %s", app.ID)
 		// Worker is deploying, leave him alone
 		if app.TasksRunning == 0 {
+			log.Debug("killAwolWorkers> app %s is deploying, do nothing", app.ID)
 			continue
 		}
 		t, err := time.Parse(time.RFC3339, app.Version)
 		if err != nil {
-			log.Warning("Cannot parse last update: %s", err)
+			log.Warning("killAwolWorkers> app %s - Cannot parse last update: %s", app.ID, err)
```
Comment (sguiheux) [inline engine/hatchery/marathon/marathon.go]: %d

File: engine/hatchery/marathon/marathon.go [hunk 23]
```diff
@@ -382,15 +388,17 @@ func (m *HatcheryMarathon) killAwolWorkers() error {
 		for _, w := range workers {
 			if strings.HasSuffix(app.ID, w.Name) && w.Status != sdk.StatusDisabled {
 				found = true
+				log.Debug("killAwolWorkers> apps %s is found on workers list with status %s", app.ID, w.Status)
```
Comment (sguiheux) [inline engine/hatchery/marathon/marathon.go]: %d

File: engine/hatchery/marathon/marathon.go [hunk 24]
```diff
@@ -382,15 +388,17 @@ func (m *HatcheryMarathon) killAwolWorkers() error {
...
 				break
 			}
 		}
 
 		// then if it's not found, kill it !
 		if !found && time.Since(t) > 1*time.Minute {
-			log.Notice("killing awol worker %s", app.ID)
+			log.Notice("killAwolWorkers> killing awol worker %s", app.ID)
```
Comment (sguiheux) [inline engine/hatchery/marathon/marathon.go]: %d

File: engine/hatchery/marathon/marathon.go [hunk 25]
```diff
@@ -382,15 +388,17 @@ func (m *HatcheryMarathon) killAwolWorkers() error {
...
 			if _, err := m.client.DeleteApplication(app.ID, true); err != nil {
-				return err
+				log.Warning("killAwolWorkers> Error while delete app %s err:%s", app.ID, err)
```
Comment (sguiheux) [inline engine/hatchery/marathon/marathon.go]: %d

File: engine/hatchery/marathon/marathon.go (hunk 22 above)
Comment (yesnault) [inline engine/hatchery/marathon/marathon.go]: nope.
```go
 ID                         string               `json:"id,omitempty"`
```

File: engine/hatchery/marathon/marathon.go (hunk 23 above)
Comment (yesnault) [inline engine/hatchery/marathon/marathon.go]: nope.
```go
 ID                         string               `json:"id,omitempty"`
```

File: engine/hatchery/marathon/marathon.go (hunk 24 above)
Comment (yesnault) [inline engine/hatchery/marathon/marathon.go]: nope.
```go
 ID                         string               `json:"id,omitempty"`
```

File: engine/hatchery/marathon/marathon.go (hunk 25 above)
Comment (yesnault) [inline engine/hatchery/marathon/marathon.go]: nope.
```go
 ID                         string               `json:"id,omitempty"`
```

**Pull request merged**